### PR TITLE
[Build] Add module cache path inside .build for swift modules

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
Fixes the inconsistent --clean behavior reported https://bugs.swift.org/browse/SR-1383

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

File: Sources/Build/Command.compile(SwiftModule).swift [hunk 1]
```diff
@@ -14,8 +14,12 @@ import Utility
...
     static func compile(swiftModule module: SwiftModule, configuration conf: Configuration, prefix: String, otherArgs: [String], SWIFT_EXEC: String) throws -> (Command, [Command]) {
 
-        let otherArgs = otherArgs + module.XccFlags(prefix) + (try module.pkgConfigArgs())
-        
+        var otherArgs = otherArgs + module.XccFlags(prefix) + (try module.pkgConfigArgs())
+
+        // Keep clang module cache path inside the build folder so cache clears on --clean.
+        let moduleCache = Path.join(prefix, "ModuleCache")
+        otherArgs += ["-module-cache-path", moduleCache]
+
```
Comment (ddunbar) [inline Sources/Build/Command.compile(SwiftModule).swift]: We should use the same module cache path for C language targets as well.

Comment (aciidb0mb3r) [inline Sources/Build/Command.compile(SwiftModule).swift]: ah yes

File: Sources/Build/misc.swift [hunk 2]
```diff
@@ -208,3 +208,27 @@ extension SystemPackageProvider {
...
+    }
+}
+
+extension SwiftModule: ClangModuleCachable {
+    func moduleCacheArgs(prefix: String) -> [String] {
+        let moduleCachePath = moduleCacheDir(prefix: prefix)
+        return ["-module-cache-path", moduleCachePath]
+    }
+}
+
```
Comment (ddunbar) [inline Sources/Build/misc.swift]: Why not put this extension on Module?

Comment (aciidb0mb3r) [inline Sources/Build/misc.swift]: I thought not all "Modules" would have this

**Comment by ddunbar** [comment]
LGTM, but we realized this has a big impact on the FunctionalTests performance... lets figure that out alongside this.

**Comment by aciidb0mb3r** [comment]
@ddunbar Added special env to handle the perf. 
@mxcl Any better suggestions?

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by ddunbar** [comment]
Chatted with Max on this, good to merge once tests pass.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test linux

**Comment by aciidb0mb3r** [comment]
@swift-ci please test linux

**Comment by aciidb0mb3r** [comment]
@swift-ci please test linux

**Comment by aciidb0mb3r** [comment]
one or other unrelated test keeps failing >_> I am going to try once again
@swift-ci please test linux

**Pull request merged**